File: python/dynamodb_service.py
# aioboto3
from utils import *
from common import Schedule, ScheduleQueryResponse
from metrics import Metrics
import json
import asyncio
import aioboto3
from boto3.dynamodb.conditions import Key

class DynamoDBService:
  def __init__(self, metrics):
    self.metrics = metrics
    self.shard_number = 10
    self.query_limit = 5
    self.table_name = 'deali_schedules'
    session = aioboto3.Session()
    self.ddb_client = session.resource('dynamodb', endpoint_url='http://localhost:8000')
  
  # def _get_dynamodb_session(self):
  #   session = aioboto3.Session()
  #   return session.resource('dynamodb', endpoint_url='http://localhost:8000')

  async def add_job(self, table) -> None:
    schedule = make_sample_schedule(self.shard_number-1)
    await table.put_item(
      Item={
        'shard_id': str(schedule.shard_id),
        'date_token': schedule.date_token,
        'job_status': schedule.job_status,
        'job_spec': json.dumps(schedule.job_spec)
      }
    )

  async def get_overdue_jobs(self, table, partition: int) -> ScheduleQueryResponse:
    start = time.time() * 1000
    now = str(start)
    response = await table.query(
      KeyConditionExpression=Key('shard_id').eq(str(partition)) & Key('date_token').lt(now),
      FilterExpression=Key('job_status').eq('SCHEDULED'),
      Limit=self.query_limit
    )
    self.metrics.get_overdue_jobs(time.time() * 1000 - start, partition)

    return ScheduleQueryResponse(
      [Schedule.decode_to_schedule(ddb_item) for ddb_item in response['Items']],
      response['Count'] == self.query_limit or 'LastEvaluatedKey' in response
    )

  async def update_status(self, table, schedule: Schedule, old_status: str, new_status: str) -> Schedule:
    response = await table.update_item(
      Key={
        'shard_id': str(schedule.shard_id), 
        'date_token': schedule.date_token
      },
      ConditionExpression='job_status = :oldStatus',
      UpdateExpression='SET job_status = :newStatus',
      ExpressionAttributeValues={
        ':oldStatus': old_status,
        ':newStatus': new_status
      },
      ReturnValues='UPDATED_NEW'
    )
    updated_status = response['Attributes']['job_status']
    schedule.job_status = updated_status
    return schedule


  async def delete_dispatched_job(self, table, schedule: Schedule) -> None:
    await table.delete_item(
      Key={
        'shard_id': str(schedule.shard_id), 
        'date_token': schedule.date_token
      },
      ConditionExpression='job_status = :acquired',
      ExpressionAttributeValues={
        ':acquired': 'ACQUIRED'
      }
    )
    
      

File: ruby/dynamodb_service.rb
require 'concurrent'
require 'aws-sdk-dynamodb'
require 'attr_extras'

class Schedule
  aattr_initialize [:shard_id!, :date_token!, :job_status!, :job_spec!]

  def self.decode_to_schedule(ddb_item)
    return self.new(
      shard_id: ddb_item['shard_id'],
      date_token: ddb_item['date_token'],
      job_status: ddb_item['job_status'],
      job_spec: JSON.parse(ddb_item['job_spec'])
    )
  end
end

class DynamoDBService
  include Concurrent::Promises::FactoryMethods

  def initialize(metrics)
    @metrics = metrics
    @dynamodb = Aws::DynamoDB::Client.new(endpoint: 'http://localhost:8000')
    @table_name = 'deali_schedules'
    @query_limit = 5
  end
  
  # return value : future<hash>
  def get_overdue_jobs(partition)
    start = Time.now
    now = (start.to_f * 1000).to_i
    return future (partition) do |partition|
      res = @dynamodb.query({
        table_name: @table_name,
        key_condition_expression: 'shard_id = :shardId AND date_token < :dateToken',
        filter_expression: 'job_status = :jobStatus',
        expression_attribute_values: {
          ':shardId': partition.to_s,
          ':dateToken': now.to_s,
          ':jobStatus': 'SCHEDULED'
        },
        limit: @query_limit
      })
      @metrics.get_overdue_jobs(Time.now-start, partition)
      {
        schedules: res.items.map { |item| Schedule.decode_to_schedule(item) },
        should_immediately_query_again: res.count == @query_limit || !!res.last_evaluated_key
      }
    end
  end

  
  # return value : future<Schedule>
  def update_status(schedule, old_status, new_status)
    return future(schedule, old_status, new_status) do |schedule, old_status, new_status|
      res = @dynamodb.update_item({
        table_name: @table_name,
        key: {
          shard_id: schedule.shard_id,
          date_token: schedule.date_token
        },
        condition_expression: 'job_status = :oldStatus',
        expression_attribute_values: {
          ':oldStatus': old_status,
          ':newStatus': new_status,
        },
        update_expression: 'SET job_status = :newStatus',
        return_values: 'UPDATED_NEW'
      })
      res
    end
    .then do |res|
      schedule.job_status = res.attributes['job_status']
      schedule
    end
  end

  # return value : future<nil>
  def delete_dispatched_jobs(schedule)
    return future(schedule) do |schedule|
      @dynamodb.delete_item({
        table_name: @table_name,
        key: {
          shard_id: schedule.shard_id,
          date_token: schedule.date_token
        },
        condition_expression: 'job_status = :acquired',
        expression_attribute_values: {
          ':acquired': 'ACQUIRED'
        }
      })
    end
  end
end
